plugin/src/main/java/org/opensearch/ml/action/models/DeleteModelTransportAction.java

```diff
@@ -84,25 +100,145 @@ public DeleteModelTransportAction(
         Settings settings,
         NamedXContentRegistry xContentRegistry,
         ClusterService clusterService,
-        ModelAccessControlHelper modelAccessControlHelper
+        ModelAccessControlHelper modelAccessControlHelper,
+        RelatedModelIdHelper relatedModelIdHelper
     ) {
         super(MLModelDeleteAction.NAME, transportService, actionFilters, MLModelDeleteRequest::new);
         this.client = client;
         this.xContentRegistry = xContentRegistry;
         this.clusterService = clusterService;
         this.modelAccessControlHelper = modelAccessControlHelper;
+        this.relatedModelIdHelper = relatedModelIdHelper;
     }
 
     @Override
     protected void doExecute(Task task, ActionRequest request, ActionListener<DeleteResponse> actionListener) {
         MLModelDeleteRequest mlModelDeleteRequest = MLModelDeleteRequest.fromActionRequest(request);
         String modelId = mlModelDeleteRequest.getModelId();
+        try (ThreadContext.StoredContext context = client.threadPool().getThreadContext().stashContext()) {
```

Comment:
Let's move the search agent models logic to a centralized place e.g. AgentModelsSearcher, and encapsulate them to return a boolean value to identify if anywhere using the model id.

Reply:
Since now we're using parallel way to check agent/pipelines, do you think it's better to put them somewhere else? For me, these logic is tightly coupled with delete action so I prefer to keep them here.

Reply:
@zane-neo Is there any other use case for AgentModelsSearcher? If not, I'm with @xinyual

Reply:
@zane-neo If so, do you think it would be better to refactor `findRelevantPipelines` into a separate class?

Reply:
My suggestion is to move all the searches to a separate class including agent search, pipeline search, but it's not mandatory as my explanation above.